### DISCUSSION TITLE
bug for "RandCropByLabelClassesd" on multiple datapoints for each batch

### DISCUSSION THREAD

**Discussion by DorisxinDU** [Project-MONAI/MONAI#6015] [General]
**Describe the bug**
 i am trying to use RandCropByLabelClassesd for image and label with following shape, but shows below bug. please can you help me. thanks.  

(8, 1, 224, 224, 64) (8, 9, 224, 224, 64)
(8, 1, 224, 224, 64) (8, 9, 224, 224, 64)
(8, 1, 224, 224, 64) (8, 9, 224, 224, 64)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [110], line 12
      8 # data=resizer(data)
      9 # data=resizer2(data)
     10 # data=reshape(data)
     11 print(data['image'].shape,data['label'].shape)
---> 12 result = cropper(data)
     13 print(result)

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/croppad/dictionary.py:999, in RandCropByLabelClassesd.__call__(self, data)
    996 image = d[self.image_key] if self.image_key else None
    997 indices = d.pop(self.indices_key, None) if self.indices_key is not None else None
--> 999 self.randomize(label, indices, image)
   1001 # initialize returned list with shallow copy to preserve key ordering
   1002 ret: List = [dict(d) for _ in range(self.cropper.num_samples)]

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/croppad/dictionary.py:991, in RandCropByLabelClassesd.randomize(self, label, indices, image)
    988 def randomize(
    989     self, label: torch.Tensor, indices: Optional[List[NdarrayOrTensor]] = None, image: Optional[torch.Tensor] = None
    990 ) -> None:
--> 991     self.cropper.randomize(label=label, indices=indices, image=image)

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/croppad/array.py:1219, in RandCropByLabelClasses.randomize(self, label, indices, image)
   1217 else:
   1218     indices_ = indices
-> 1219 self.centers = generate_label_classes_crop_centers(
   1220     self.spatial_size, self.num_samples, label.shape[1:], indices_, self.ratios, self.R, self.allow_smaller
   1221 )

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/utils.py:558, in generate_label_classes_crop_centers(spatial_size, num_samples, label_spatial_shape, indices, ratios, rand_state, allow_smaller)
    556 ratios_: List[Union[float, int]] = ([1] * len(indices)) if ratios is None else ratios
    557 if len(ratios_) != len(indices):
--> 558     raise ValueError("random crop ratios must match the number of indices of classes.")
    559 if any(i < 0 for i in ratios_):
    560     raise ValueError("ratios should not contain negative number.")

ValueError: random crop ratios must match the number of indices of classes.

**Comment by KumoLiu** [comment]
Hi @DorisxinDU, from the error message, it looks like you didn't specify the length of the ratios or indices correctly.
Could you please share how you instantiate `RandCropByLabelClassesd`? Thanks!

**Comment by DorisxinDU** [comment]
Hi @KumoLiu , thanks for reply.
The definition is: cropper=RandCropByLabelClassesd(keys=['image','label'],label_key='label_stacked',num_classes=10,spatial_size=roi_size,num_samples=4) 
result = cropper(data)
where data={'image': datapoints['image'],'label':datapoints['label'],"label_stacked":datapoints["label"]}, and the shape of datapoints['image'] is (8, 1, 224, 224, 64), the shape of datapoints["label"] is (8, 9, 224, 224, 64). the order of the shape is (Batch,channel,width,length, num_slides). I tried both using binary one-hot datapoints["label"]  and class based datapoints["label"] by changing it to (8, 1, 224, 224, 64). However both cases do not work. 

Please see the following error using classes based error:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [98], line 20
     10 cropper=RandCropByLabelClassesd(keys=['image','label'],label_key='label_stacked',num_classes=10,spatial_size=roi_size,num_samples=4)
     11 # data={'image': datapoints['image'][0],'label':datapoints['label'][0]}     ##remove [0] will not work
     12 # print(data['image'].shape,data['label'].shape)
     13 # # reshape=EnsureChannelFirstd(keys=["image"])
   (...)
     18 # # data=reshape(data)
     19 # print(data['image'].shape,data['label'].shape)
---> 20 result = cropper(data)

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/croppad/dictionary.py:999, in RandCropByLabelClassesd.__call__(self, data)
    996 image = d[self.image_key] if self.image_key else None
    997 indices = d.pop(self.indices_key, None) if self.indices_key is not None else None
--> 999 self.randomize(label, indices, image)
   1001 # initialize returned list with shallow copy to preserve key ordering
   1002 ret: List = [dict(d) for _ in range(self.cropper.num_samples)]

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/croppad/dictionary.py:991, in RandCropByLabelClassesd.randomize(self, label, indices, image)
    988 def randomize(
    989     self, label: torch.Tensor, indices: Optional[List[NdarrayOrTensor]] = None, image: Optional[torch.Tensor] = None
    990 ) -> None:
--> 991     self.cropper.randomize(label=label, indices=indices, image=image)

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/croppad/array.py:1219, in RandCropByLabelClasses.randomize(self, label, indices, image)
   1217 else:
   1218     indices_ = indices
-> 1219 self.centers = generate_label_classes_crop_centers(
   1220     self.spatial_size, self.num_samples, label.shape[1:], indices_, self.ratios, self.R, self.allow_smaller
   1221 )

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/utils.py:575, in generate_label_classes_crop_centers(spatial_size, num_samples, label_spatial_shape, indices, ratios, rand_state, allow_smaller)
    573     center = unravel_index(indices_to_use[random_int], label_spatial_shape).tolist()
    574     # shift center to range of valid centers
--> 575     centers.append(correct_crop_centers(center, spatial_size, label_spatial_shape, allow_smaller))
    577 return centers

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/transforms/utils.py:444, in correct_crop_centers(centers, spatial_size, label_spatial_shape, allow_smaller)
    426 def correct_crop_centers(
    427     centers: List[int],
    428     spatial_size: Union[Sequence[int], int],
    429     label_spatial_shape: Sequence[int],
    430     allow_smaller: bool = False,
    431 ):
    432     """
    433     Utility to correct the crop center if the crop size and centers are not compatible with the image size.
    434 
   (...)
    442 
    443     """
--> 444     spatial_size = fall_back_tuple(spatial_size, default=label_spatial_shape)
    445     if any(np.subtract(label_spatial_shape, spatial_size) < 0):
    446         if not allow_smaller:

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/utils/misc.py:201, in fall_back_tuple(user_provided, default, func)
    164 """
    165 Refine `user_provided` according to the `default`, and returns as a validated tuple.
    166 
   (...)
    198 
    199 """
    200 ndim = len(default)
--> 201 user = ensure_tuple_rep(user_provided, ndim)
    202 return tuple(  # use the default values if user provided is not valid
    203     user_c if func(user_c) else default_c for default_c, user_c in zip(default, user)
    204 )

File ~/anaconda3/envs/transform/lib/python3.9/site-packages/monai/utils/misc.py:158, in ensure_tuple_rep(tup, dim)
    155 if len(tup) == dim:
    156     return tuple(tup)
--> 158 raise ValueError(f"Sequence must have length {dim}, got {len(tup)}.")

ValueError: Sequence must have length 4, got 3.

**Comment by KumoLiu** [comment]
Hi @DorisxinDU, I see your questions. I think 8 is your batch size. The images and labels should be channel-first.
https://github.com/Project-MONAI/MONAI/blob/11745a6f7683effd40ff55f88306d4463f9d66a9/monai/transforms/croppad/array.py#L1236-L1237

```
from monai.transforms import RandCropByLabelClassesd
import numpy as np

cropper = RandCropByLabelClassesd(
    keys=["image", "label"],
    label_key="label",
    spatial_size=[3, 3],
    # ratios=[1, 2, 3, 1],   # optional
    num_classes=4,
    num_samples=2,
)
data = {
    "image": np.array([
        [[0.0, 0.3, 0.4, 0.2, 0.0],
        [0.0, 0.1, 0.2, 0.1, 0.4],
        [0.0, 0.3, 0.5, 0.2, 0.0],
        [0.1, 0.2, 0.1, 0.1, 0.0],
        [0.0, 0.1, 0.2, 0.1, 0.0]]
    ])[None],
    "label": np.array([
        [[0, 0, 0, 0, 0],
        [0, 1, 2, 1, 0],
        [0, 1, 3, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]]
    ])[None],
}
result = cropper(data)
```

Hope it can help you, thanks! Just convert it to discussions.
